src/centrality.rs

```diff
@@ -0,0 +1,296 @@
+                betweenness[i] = Some(betweenness[i].unwrap() * scale);
+            }
+        }
+    }
+}
+
+fn _accumulate_basic(
+    betweenness: &mut Vec<Option<f64>>,
+    max_index: usize,
+    path_calc: &mut ShortestPathData,
+    is: usize,
+) {
+    let mut delta = vec![0.0; max_index];
+    for w in &path_calc.verts_sorted_by_distance {
+        let iw = w.index();
+        let coeff = (1.0 + delta[iw]) / (path_calc.sigma[&iw] as f64);
+        let p_w = path_calc.predecessors.get_mut(w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (path_calc.sigma[&iv] as f64) * coeff;
+        }
+        if iw != is && betweenness[iw].is_some() {
+            betweenness[iw] = Some(betweenness[iw].unwrap() + delta[iw]);
+        }
+    }
+}
+
+fn _accumulate_endpoints(
+    betweenness: &mut Vec<Option<f64>>,
+    max_index: usize,
+    path_calc: &mut ShortestPathData,
+    is: usize,
+) {
+    if betweenness[is].is_some() {
+        betweenness[is] = Some(
+            betweenness[is].unwrap()
+                + ((path_calc.verts_sorted_by_distance.len() - 1) as f64),
+        );
+    }
+    let mut delta = vec![0.0; max_index];
+    for w in &path_calc.verts_sorted_by_distance {
+        let iw = w.index();
+        let coeff = (1.0 + delta[iw]) / (path_calc.sigma[&iw] as f64);
+        let p_w = path_calc.predecessors.get_mut(w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (path_calc.sigma[&iv] as f64) * coeff;
+        }
+        if iw != is && betweenness[iw].is_some() {
+            betweenness[iw] = Some(betweenness[iw].unwrap() + delta[iw] + 1.0);
+        }
+    }
+}
+
+struct ShortestPathData {
+    verts_sorted_by_distance: Vec<NodeIndex>,
+    predecessors: HashMap<NodeIndex, Vec<NodeIndex>>,
+    sigma: HashMap<usize, i64>,
+}
+
+fn shortest_path_for_centrality<G>(
+    graph: G,
+    node_s: &G::NodeId,
+) -> ShortestPathData
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + GraphBase<NodeId = NodeIndex>, // for get() and get_mut()
+{
+    let mut verts_sorted_by_distance: Vec<NodeIndex> = Vec::new(); // a stack
+    #[allow(non_snake_case)]
+    let mut Q: Vec<NodeIndex> = Vec::new(); // a queue
+    let mut predecessors = HashMap::<G::NodeId, Vec<G::NodeId>>::new();
+    let mut sigma = HashMap::<usize, i64>::new();
+    let mut distance = HashMap::<G::NodeId, i64>::new();
+
+    let i_s = graph.to_index(*node_s);
+    let index_s = NodeIndex::new(i_s);
+
+    for node in graph.node_identifiers() {
+        let i = graph.to_index(node);
+        let index = NodeIndex::new(i);
+        predecessors.insert(index, Vec::new());
+        sigma.insert(i, 0);
+        distance.insert(index, -1);
+    }
+    sigma.insert(index_s.index(), 1);
+    distance.insert(index_s, 0);
+    Q.push(index_s);
+    while !Q.is_empty() {
+        let v = Q.remove(0);
```

Comment:
If we're using `Q` as a FIFO I'm wondering if makes more sense to use a `VecDeque` here https://doc.rust-lang.org/std/collections/struct.VecDeque.html because doing `Vec.remove(0)` requires shifting every element in the queue  while `VecDeque` is a ring buffer and avoids that.